### DISCUSSION TITLE
Strange failure while testing plumber API call when using docker images

### DISCUSSION THREAD

**Discussion by rhuettl** [revodavid/R-custom-handler#3] [General]
Hi David, hi all,

Thank you for the sample and ideas. I'm just facing a strange problem when I build the container and test locally (a test via func start --custom and a test in R is working fine without issues). 

I tried all kind of things but I don't find the problem. Maybe it has something to do that the port or request forwarding to plumber is not working as expected:

![2020-12-22_5-23-00](https://user-images.githubusercontent.com/23125000/102849052-8e29b780-4416-11eb-8efc-ea965ef9f7a5.jpg)

Calling the function in a browser is giving a 404 result back (but as mentioned before, just within the docker test, all other tests without docker are working successful):
![image](https://user-images.githubusercontent.com/23125000/102849250-07290f00-4417-11eb-8c6b-4c2d7e56e020.png)

And the log is then looking like:
![image](https://user-images.githubusercontent.com/23125000/102849152-d1842600-4416-11eb-8c31-d35beb7172b1.png)

Here's the trace log:
![image](https://user-images.githubusercontent.com/23125000/102853813-5c6a1e00-4421-11eb-9b7e-9d39871c8779.png)

So the Azure function has been called but not the plumber function as itself. Looking like the plumber API / request isn't coming through...

Any idea would be helpful :-) 

Thank you in advance, 
Ralf

**Comment by revodavid** [comment]
I haven't seen an error like this before. Have you tried deploying to Functions and does the container work there? Or does the error only happen when the container is run locally?

**Comment by revodavid** [comment]
Another thing you can try is to use the `message()` function in R to trace execution in the log file above. Does your `launch-service.R` file get run? How about `handler.R`?

**Reply by rhuettl** [reply]
The launch-service.R gets called (but i don't know why it is marked as failed):
![image](https://user-images.githubusercontent.com/23125000/102938829-5588ed00-44ad-11eb-8ea8-a2b8e2ace77e.png)

![image](https://user-images.githubusercontent.com/23125000/102938773-3f7b2c80-44ad-11eb-994f-6641493183de.png)

Seems that the request won't be forwarded to plumber API because when calling the msg function there is no log:
![image](https://user-images.githubusercontent.com/23125000/102938910-836e3180-44ad-11eb-8836-fb5cb8e6cd99.png)
I just can see that the function has been called but not forwarded to the plumber implementation/API. Same effect in your example. So there must be a problem withing the communication between plumber and the HTTP Worker Channel. I'm not able to see more details in the trace logs, what's going on.

Calling the request without the container (by "func start") it's all working fine. Just the combination of function and container is making some troble at this place :-(

**Reply by rhuettl** [reply]
One update. I deployed the container and function to Azure but it seems the same problem, timing out:
![image](https://user-images.githubusercontent.com/23125000/102943120-75bda980-44b7-11eb-9a05-c4349b625937.png)

But in my browser it's still loading the request a while, then ending in HTTP 504 (nginx Gaeteway Timeout)... very strange behavior.

Really confusing me :-) Because on the function monitoring I also can't see any executions:
![image](https://user-images.githubusercontent.com/23125000/102943428-4ce9e400-44b8-11eb-9b79-e91ea8a8e290.png)

Trying the same with your sources is resulting in the same bahavior, also giving the strange timeouts / forward request problems. I'm not able to see any failure in our configuration or scripts. Seems to be an internal forwarding problem, maybe you find out some more because it's a cool feature and we're really interested in using that for scaling our R applications in a more comfortbale way.

Oh, one final thing I tried was replacing plumber with httpuv (tried exactly with the example provided via https://docs.microsoft.com/en-us/azure/azure-functions/functions-create-function-linux-custom-image?pivots=programming-language-other&tabs=bash%2Cportal&WT.mc_id=javascript-10496-davidsmi). Also resulting in a 404 when running in a container.

Azure CLI used 2.16.0
Docker on Windows 10 used 20.10.0
Function Core used 3.0.3160


Giving it up for today :-)

Greetings, merry Christmas
Ralf

PS. I uploaded my example at https://github.com/rhuettl/azurefuncR

**Reply by anthonychu** [reply]
Just catching up on this coming back from vacation. Can you run the handler by itself and hit /api/msg with a GET request?

**Comment by revodavid** [comment]
Looks like a networking issue for me, but I'm not sure how to solve it. @anthonychu might be able to help.

**Reply by revodavid** [reply]
By the way, are you running Docker and the CLI from Windows directly, or via WSL? If WSL, is it the original WSL or WSL2?

**Reply by rhuettl** [reply]
I'm running Docker and CLI from Windows command line directly. I also think that a network setting (either in the Function Host or Docker configuration) is the problem. Also no more ideas at the moment :-( I also will contact https://github.com/anthonychu. We like to use this amazing new feature but I'm curretnly a bit frustrated :-) 
Ralf

**Reply by rhuettl** [reply]
Maybe you find a few minutes for just creating the Docker image and run it locally on your side (https://github.com/rhuettl/azurefuncR). If this isn't also working then I think the Function Host is having a problem :-) Let's see.

**Reply by rhuettl** [reply]
My final update for today :-) 

I re-tested the docker built on a personal/private server and not on a company one and the call succeeded!
Seems that docker is creating or starting the image different this way - but running the same version, really strange.

Ralf

**Reply by revodavid** [reply]
That's great to hear, thanks for sharing. Maybe there are IT policies that interfere with Docker on Windows?

**Comment by javianpicardoarchi** [comment]
I am having a similar Timeout issue with Azure function running the plumber --->> After running the docker run command
docker build --tag javianpicardoarchi/rcoefficientazfuncimg:v1.0.0 .
docker run -p 8080:80 -it javianpicardoarchi/rcoefficientazfuncimg:v1.0.0

I have added the screenshot to show the dockerfile & host.json code.
https://github.com/ArchiMedia-Advantage-Inc-Internal/RCoefficientProject.git
Any help will be appreciated. 
@rhuettl @revodavid @anthonychu 

![image](https://user-images.githubusercontent.com/82883817/124180575-60d35b80-da82-11eb-994d-a81585345342.png)

**Reply by javianpicardoarchi** [reply]
Here I have pointed to 8080 port but also tried with the PORTEnv <- Sys.getenv("FUNCTIONS_CUSTOMHANDLER_PORT") but it gives the same error.

**Reply by javianpicardoarchi** [reply]
Git Code Path for reference : https://github.com/ArchiMedia-Advantage-Inc-Internal/RCoefficientProject.git

**Comment by wjchristenson2** [comment]
I also am having the same issue. The function runtime is not able to communicate w/ or forward requests to the customerHandler, nor can I manually hit the plumber api endpoint via a browser directly. I connected to the docker container terminal window to manually start and wire up plumber. That went ok without error. It runs on port 8080, but no requests seem to go through/hit it. They 404.

> I just can see that the function has been called but not forwarded to the plumber implementation/API. Same effect in your example. So there must be a problem withing the communication between plumber and the HTTP Worker Channel. I'm not able to see more details in the trace logs, what's going on.

Same as rhuettl. I can see the http request come in via the trace, but then 404. I've tried different Microsoft.Azure.Functions.ExtensionBundle versions, different azure function base image versions to no avail.
